neard/src/genesis_validate.rs

```diff
@@ -0,0 +1,189 @@
+                }
+                total_supply += account.locked + account.amount;
+                account_ids.insert(account_id.clone());
+                if account.locked > 0 {
+                    staked_accounts.insert(account_id.clone(), account.locked);
+                }
+            }
+            StateRecord::AccessKey { account_id, .. } => {
+                access_key_account_ids.insert(account_id.clone());
+            }
+            StateRecord::Contract { account_id, .. } => {
+                if contract_account_ids.contains(account_id) {
+                    panic!("account {} has more than one contract deployed", account_id);
+                }
+                contract_account_ids.insert(account_id.clone());
+            }
+            _ => {}
+        }
+    }
+    assert_eq!(total_supply, genesis.config.total_supply, "wrong total supply");
+    assert_eq!(validators, staked_accounts, "validator accounts do not match staked accounts");
+    for account_id in access_key_account_ids {
+        assert!(
+            account_ids.contains(&account_id),
+            "access key account {} does not exist",
+            account_id
+        );
+    }
+    for account_id in contract_account_ids {
+        assert!(
+            account_ids.contains(&account_id),
+            "contract account {} does not exist",
+            account_id
+        );
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::genesis_validate::validate_genesis;
+    use near_chain_configs::{Genesis, GenesisRecords};
+    use near_crypto::{KeyType, PublicKey};
+    use near_primitives::serialize::to_base64;
+    use near_primitives::state_record::StateRecord;
+    use near_primitives::types::AccountInfo;
+    use near_primitives::views::{AccessKeyPermissionView, AccessKeyView, AccountView};
+
+    #[test]
+    #[should_panic(expected = "wrong total supply")]
+    fn test_total_supply_not_match() {
+        let mut genesis = Genesis::default();
+        genesis.config.validators = vec![AccountInfo {
+            account_id: "test".to_string(),
+            public_key: PublicKey::empty(KeyType::ED25519),
+            amount: 10,
+        }];
+        genesis.records = GenesisRecords(vec![StateRecord::Account {
+            account_id: "test".to_string(),
+            account: AccountView {
+                amount: 100,
+                locked: 10,
+                code_hash: Default::default(),
+                storage_usage: 0,
+                storage_paid_at: 0,
+            },
+        }]);
+        validate_genesis(&genesis);
+    }
+
+    #[test]
+    #[should_panic(expected = "validator accounts do not match staked accounts")]
+    fn test_validator_not_match() {
+        let mut genesis = Genesis::default();
+        genesis.config.validators = vec![AccountInfo {
+            account_id: "test".to_string(),
+            public_key: PublicKey::empty(KeyType::ED25519),
+            amount: 100,
+        }];
+        genesis.config.total_supply = 110;
```

Comment:
Don't we compute total_supply when we load genesis from records in the first place?

Reply:
doesn't hurt to double check